File: apps/server/src/adapter/driven/cloudinary/storage.ts
import { v2 as cloudinary, UploadApiResponse } from "cloudinary";
import { Readable } from "node:stream";
import type { FileStorage } from "../../../domain/port/index.js";

export interface CloudinaryStorageConfig {
  cloudName: string;
  apiKey: string;
  apiSecret: string;
}

export class CloudinaryFileStorage implements FileStorage {
  constructor(config: CloudinaryStorageConfig) {
    cloudinary.config({
      cloud_name: config.cloudName,
      api_key: config.apiKey,
      api_secret: config.apiSecret,
      secure: true,
    });
  }

  async upload(
    file: Buffer,
    fileName: string,
    _mimeType: string,
  ): Promise<{ url: string; publicId: string }> {
    const result = await new Promise<UploadApiResponse>((resolve, reject) => {
      const stream = cloudinary.uploader.upload_stream(
        {
          public_id: fileName.replace(/\.[^.]+$/, ""),
          resource_type: "auto",
          folder: "neurodyne",
        },
        (error, result) => {
          if (error) return reject(error);
          resolve(result!);
        },
      );

      const readable = Readable.from(file);
      readable.pipe(stream);
    });

    return {
      url: result.secure_url,
      publicId: result.public_id,
    };
  }

  async delete(publicId: string): Promise<void> {
    await cloudinary.uploader.destroy(publicId);
  }

  getURL(publicId: string): string {
    return cloudinary.url(publicId, { secure: true });
  }
}

File: apps/api/internal/adapter/driven/cloudinary/storage.go
package cloudinary

import (
	"context"
	"io"

	"github.com/cloudinary/cloudinary-go/v2"
	"github.com/cloudinary/cloudinary-go/v2/api/uploader"
	"github.com/neurodynecorp/api/internal/domain/port"
	"github.com/neurodynecorp/api/pkg/logger"
)

type Storage struct {
	client *cloudinary.Cloudinary
	log    *logger.Logger
}

func NewStorage(cloudinaryURL string, log *logger.Logger) (*Storage, error) {
	l := log.WithComponent("cloudinary")

	cld, err := cloudinary.NewFromURL(cloudinaryURL)
	if err != nil {
		l.Error().Err(err).Msg("failed to initialize Cloudinary")
		return nil, err
	}

	l.Info().Msg("Cloudinary storage initialized")

	return &Storage{
		client: cld,
		log:    l,
	}, nil
}

func (s *Storage) Upload(ctx context.Context, file io.Reader, filename string, folder string) (string, error) {
	result, err := s.client.Upload.Upload(ctx, file, uploader.UploadParams{
		PublicID: filename,
		Folder:   "neurodyne/" + folder,
	})
	if err != nil {
		s.log.Error().Err(err).Str("filename", filename).Msg("upload failed")
		return "", err
	}

	s.log.Debug().Str("public_id", result.PublicID).Msg("file uploaded")
	return result.SecureURL, nil
}

func (s *Storage) Delete(ctx context.Context, publicID string) error {
	_, err := s.client.Upload.Destroy(ctx, uploader.DestroyParams{
		PublicID: publicID,
	})
	return err
}

func (s *Storage) GetURL(publicID string) string {
	asset, _ := s.client.Image(publicID)
	if asset == nil {
		return ""
	}
	url, _ := asset.String()
	return url
}

var _ port.FileStorage = (*Storage)(nil)
